api/grpc/controller_test.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package grpc
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	tmock "github.com/stretchr/testify/mock"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/dps/mock"
+)
+
+func TestNewController(t *testing.T) {
+	c := NewController(nil)
+	assert.NotNil(t, c)
+}
+
+func TestController_GetRegister(t *testing.T) {
+	var (
+		testHeight uint64 = 128
+		lastHeight uint64 = 256
+
+		testKey   = []byte(`testKey`)
+		testValue = []byte(`testValue`)
+	)
+
+	tests := []struct {
+		desc string
+
+		reqHeight *uint64
+		reqKey    []byte
+
+		mockValue []byte
+		mockErr   error
+
+		wantHeight uint64
+		wantResp   *GetRegisterResponse
+		wantErr    assert.ErrorAssertionFunc
+	}{
+		{
+			desc: "nominal case, height given",
+
+			reqHeight: &testHeight,
+			reqKey:    testKey,
+
+			mockValue: testValue,
+
+			wantHeight: testHeight,
+			wantResp: &GetRegisterResponse{
+				Height: testHeight,
+				Key:    testKey,
+				Value:  testValue,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			desc: "nominal case, no height given",
+
+			reqKey: testKey,
+
+			mockValue: testValue,
+
+			wantHeight: lastHeight,
+			wantResp: &GetRegisterResponse{
+				Height: lastHeight,
+				Key:    testKey,
+				Value:  testValue,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			desc: "state error",
+
+			reqKey: testKey,
+
+			mockErr: errors.New("dummy error"),
+
+			wantHeight: lastHeight,
+			wantResp:   nil,
+			wantErr:    assert.Error,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+
+			m := mock.NewState()
+			m.LastState.On("Height").Return(lastHeight).Once()
+			m.RawState.On("WithHeight", test.wantHeight).Return(m.RawState).Once()
+			m.RawState.On("Get", test.reqKey).Return(test.mockValue, test.mockErr)
```

Comment:
Are we now using testify mocks after all? I thought we agreed not to do that because these `On` calls are not strictly typed?

Reply:
I am not fond of the non-strictly-typed `On` calls indeed but since the state mock is needed throughout multiple test packages, I thought it needs to be a mock that is easy to configure and use so as not to make tests unreadable. I think `testify/mock`'s approach, while not perfect, fits this use case.

Do you have ideas on how to achieve a similar clean result without using `testify/mock`?

Reply:
I personally don't quite understand why it's cleaner. You could just have your function definitions in place of the `On` calls here, no? I honestly preferred your earlier way. After I saw it the first time it made a lot of sense.

In the case for the state with its chainable API, you could have one field for each returned sub-struct, and just define the final functions on these, no?

Reply:
That's the way it was done before and it was a mess tbh:

```go
			stateMock := &stateMock{
				withHeight: func(height uint64) dps.Raw {
					return &stateMock{
						get: test.stateGet,
					}
				},
				last: lastMock{
					height: func() uint64 {
						return test.lastHeight
					},
				},
			}
```

I think handmade mocks are fine for simple cases but here with the chainable API you need to set the methods on whichever part will end up making the call, which means either nesting a lot of defining the mock over multiple declarations, none of which I like :/

Reply:
What about this:

```go
type stateMock struct {
  raw rawMock
  height heightMock
  last lastMock
}

func (s *stateMock) Raw() dps.Raw {
  return s.raw
}

func (s *stateMock) Height() dps.Height {
  return s.height
}

func (s *stateMock) Last() dps.Last {
  return s.last
}

type rawMock struct {
  get func(uint64, []byte) ([]byte, error)
}

type heightMock struct {
  forBlock func(flow.Identifier) (uint64, error)
  forCommit func(flow.StateCommitment) (uint64, error)
}

// ... and so on
```

Then you can program it like this:

```go
stateMock.raw.get = func(...)
// and so on ...
```